File: Solid_Principles/Liskov_Substitution_Principle/03_LSP.js
// TITLE: Liskov Substitution Principle (LSP) - Vehicle Example

// THE PROBLEM (VIOLATION):
class Vehicle {
    getNumberOfWheels() { throw new Error("Must implement"); }
    hasEngine() { throw new Error("Must implement"); } // Problem!
}

class MotorCycle extends Vehicle {
    getNumberOfWheels() { return 2; }
    hasEngine() { return true; }
}

class Car extends Vehicle {
    getNumberOfWheels() { return 4; }
    hasEngine() { return true; }
}

class Bicycle extends Vehicle {
    getNumberOfWheels() { return 2; }
    hasEngine() {
        // Violation! Bicycle doesn't have an engine
        return false; // Misleading!
    }
}

// THE SOLUTION:
class VehicleGood {
    getNumberOfWheels() { throw new Error("Must implement"); }
}

class EngineVehicle extends VehicleGood {
    hasEngine() { return true; }
}

class MotorCycleGood extends EngineVehicle {
    getNumberOfWheels() { return 2; }
}

class CarGood extends EngineVehicle {
    getNumberOfWheels() { return 4; }
}

class BicycleGood extends VehicleGood {
    getNumberOfWheels() { return 2; }
    // No hasEngine() method - doesn't apply!
}

// Usage
console.log("=== PROBLEM (LSP Violation) ===");
const bicycle = new Bicycle();
console.log(`Bicycle wheels: ${bicycle.getNumberOfWheels()}`);
console.log(`Bicycle has engine: ${bicycle.hasEngine()} (Misleading!)`);

console.log("\n=== SOLUTION (LSP Compliant) ===");
const bicycleGood = new BicycleGood();
console.log(`Bicycle wheels: ${bicycleGood.getNumberOfWheels()}`);
console.log("Bicycle doesn't need hasEngine() method!");

const car = new CarGood();
console.log(`Car wheels: ${car.getNumberOfWheels()}`);
console.log(`Car has engine: ${car.hasEngine()}`);

File: Solid_Principles/Liskov_Substitution_Principle/03_LSP.py
from abc import ABC, abstractmethod

# TITLE: Liskov Substitution Principle (LSP) - Vehicle Example
# Source: Shrayansh Jain's video (129QkkXUHeQ)

# THE PROBLEM (VIOLATION):
class Vehicle(ABC):
    @abstractmethod
    def get_number_of_wheels(self): pass
    
    @abstractmethod
    def has_engine(self): pass  # Problem: Not all vehicles have engines!

class MotorCycle(Vehicle):
    def get_number_of_wheels(self): return 2
    def has_engine(self): return True

class Car(Vehicle):
    def get_number_of_wheels(self): return 4
    def has_engine(self): return True

class Bicycle(Vehicle):
    def get_number_of_wheels(self): return 2
    def has_engine(self):
        # Violation! Bicycle doesn't have an engine
        return False  # Misleading!

# THE SOLUTION:
class VehicleGood(ABC):
    @abstractmethod
    def get_number_of_wheels(self): pass

class EngineVehicle(VehicleGood):
    def has_engine(self): return True

class MotorCycleGood(EngineVehicle):
    def get_number_of_wheels(self): return 2

class CarGood(EngineVehicle):
    def get_number_of_wheels(self): return 4

class BicycleGood(VehicleGood):
    def get_number_of_wheels(self): return 2
    # No has_engine() method - doesn't apply!

if __name__ == "__main__":
    print("=== PROBLEM (LSP Violation) ===")
    bicycle = Bicycle()
    print(f"Bicycle wheels: {bicycle.get_number_of_wheels()}")
    print(f"Bicycle has engine: {bicycle.has_engine()} (Misleading!)")
    
    print("\n=== SOLUTION (LSP Compliant) ===")
    bicycle_good = BicycleGood()
    print(f"Bicycle wheels: {bicycle_good.get_number_of_wheels()}")
    print("Bicycle doesn't need has_engine() method!")
    
    car = CarGood()
    print(f"Car wheels: {car.get_number_of_wheels()}")
    print(f"Car has engine: {car.has_engine()}")
